sse2neon.h

```diff
@@ -8358,6 +8359,526 @@ FORCE_INLINE int _mm_testz_si128(__m128i a, __m128i b)
 
 /* SSE4.2 */
 
+static uint16_t _sse2neon_cmpestr_mask16b[8] __attribute__((aligned(16))) = {
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+};
+static uint8_t _sse2neon_cmpestr_mask8b[16] __attribute__((aligned(16))) = {
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80,
+};
+
+/* specify the source data format */
+#define _SIDD_UBYTE_OPS 0x00 /* unsigned 8-bit characters */
+#define _SIDD_UWORD_OPS 0x01 /* unsigned 16-bit characters */
+#define _SIDD_SBYTE_OPS 0x02 /* signed 8-bit characters */
+#define _SIDD_SWORD_OPS 0x03 /* signed 16-bit characters */
+
+/* specify the comparison operation */
+#define _SIDD_CMP_EQUAL_ANY 0x00     /* compare equal any: strchr */
+#define _SIDD_CMP_RANGES 0x04        /* compare ranges */
+#define _SIDD_CMP_EQUAL_EACH 0x08    /* compare equal each: strcmp */
+#define _SIDD_CMP_EQUAL_ORDERED 0x0C /* compare equal ordered */
+
+/* specify the polarity */
+#define _SIDD_POSITIVE_POLARITY 0x00
+#define _SIDD_MASKED_POSITIVE_POLARITY 0x20
+#define _SIDD_NEGATIVE_POLARITY 0x10 /* negate results */
+#define _SIDD_MASKED_NEGATIVE_POLARITY \
+    0x30 /* negate results only before end of string */
+
+/* specify the output selection in _mm_cmpXstri */
+#define _SIDD_LEAST_SIGNIFICANT 0x00
+#define _SIDD_MOST_SIGNIFICANT 0x40
+
+/* specify the output selection in _mm_cmpXstrm */
+#define _SIDD_BIT_MASK 0x00
+#define _SIDD_UNIT_MASK 0x40
+
+#define EQ_U8(i)                                                          \
+    mtx[i] = vreinterpretq_m128i_u8(                                      \
+        vceqq_u8(vdupq_n_u8(vgetq_lane_u8(vreinterpretq_u8_m128i(b), i)), \
+                 vreinterpretq_u8_m128i(a)));
+#define EQ_U16(i)                                                             \
+    mtx[i] = vreinterpretq_m128i_u16(                                         \
+        vceqq_u16(vdupq_n_u16(vgetq_lane_u16(vreinterpretq_u16_m128i(b), i)), \
+                  vreinterpretq_u16_m128i(a)));
+#define RNG_U8(i) \
+    vec_b[i] = vdupq_n_u8(vgetq_lane_u8(vreinterpretq_u8_m128i(b), i));
+#define RNG_S8(i) \
+    vec_b[i] = vdupq_n_s8(vgetq_lane_s8(vreinterpretq_s8_m128i(b), i));
+#define RNG_U16(i) \
+    vec_b[i] = vdupq_n_u16(vgetq_lane_u16(vreinterpretq_u16_m128i(b), i));
+#define RNG_S16(i) \
+    vec_b[i] = vdupq_n_s16(vgetq_lane_s16(vreinterpretq_s16_m128i(b), i));
+
+#define EIGHT(f) f(0) f(1) f(2) f(3) f(4) f(5) f(6) f(7)
```

Comment:
Use the manner of implementing loops: https://stackoverflow.com/questions/28231743/self-unrolling-macro-loop-in-c-c